File: typescript/packages/core/src/commands/builtin/generic/crossmount/fanout/fanout.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { materialize, type ByteSource } from '../../../../../io/types.ts'
import type { PathSpec } from '../../../../../types.ts'
import { combinedExit } from './exit.ts'
import { duTotal } from './du.ts'
import { combineWc } from './wc.ts'
import { Cmd, type CrossResult, type OperandRun, type RunSingle } from '../types.ts'
import { mergeOperandIos, runOperands } from '../utils.ts'

const ENC = new TextEncoder()

function concatRuns(results: OperandRun[]): Uint8Array {
  const nonEmpty = results.map((r) => r.data).filter((d) => d.byteLength > 0)
  const size = nonEmpty.reduce((n, d) => n + d.byteLength, 0)
  const out = new Uint8Array(size)
  let offset = 0
  for (const d of nonEmpty) {
    out.set(d, offset)
    offset += d.byteLength
  }
  return out
}

function joinRunsWithBlankLine(results: OperandRun[]): Uint8Array {
  const parts = results.map((r) => r.data).filter((d) => d.byteLength > 0)
  const sep = ENC.encode('\n')
  const size =
    parts.reduce((n, d) => n + d.byteLength, 0) + sep.byteLength * Math.max(0, parts.length - 1)
  const out = new Uint8Array(size)
  let offset = 0
  parts.forEach((d, i) => {
    if (i > 0) {
      out.set(sep, offset)
      offset += sep.byteLength
    }
    out.set(d, offset)
    offset += d.byteLength
  })
  return out
}

// Run a per-operand command whose operands span mounts. The command runs
// natively once per operand on the operand's owning mount (globs expand
// inside that native run), and the outputs combine in operand order.
// Filename-keyed commands stay correct because every native run is forced to
// name its files (grep `-H`, head/tail `-v`); wc and `du -c` re-total across
// runs.
export async function runFanout(
  cmdName: Cmd,
  scopes: PathSpec[],
  textArgs: string[],
  flagKwargs: Record<string, string | boolean | string[]>,
  runSingle: RunSingle,
  stdin: ByteSource | null = null,
): Promise<CrossResult> {
  const flags = { ...flagKwargs }
  let stdinBytes: Uint8Array | null = null
  if (cmdName === Cmd.TEE) {
    stdinBytes = stdin !== null ? await materialize(stdin) : new Uint8Array()
  }
  if (cmdName === Cmd.GREP && flags.h !== true) {
    flags.H = true
  }
  if (cmdName === Cmd.RG && flags.args_I !== true) {
    flags.H = true
  }
  if ((cmdName === Cmd.HEAD || cmdName === Cmd.TAIL) && flags.q !== true) {
    flags.v = true
  }

  const results = await runOperands(runSingle, cmdName, scopes, [...textArgs], flags, stdinBytes)
  const exitCode = combinedExit(
    cmdName,
    results.map((r) => r.io.exitCode),
  )

  let body: Uint8Array
  if (cmdName === Cmd.WC) {
    body = combineWc(results, flagKwargs)
  } else if (cmdName === Cmd.DU && flagKwargs.c === true) {
    body = duTotal(results, flagKwargs.h === true)
  } else if (cmdName === Cmd.TEE) {
    body = stdinBytes ?? new Uint8Array()
  } else if (
    ((cmdName === Cmd.HEAD || cmdName === Cmd.TAIL) && flags.v === true) ||
    (cmdName === Cmd.LS && flagKwargs.R === true)
  ) {
    // Blank line between per-operand blocks, like one native run separates
    // its own file blocks.
    body = joinRunsWithBlankLine(results)
  } else {
    body = concatRuns(results)
  }

  const io = await mergeOperandIos(results, exitCode)
  return [body, io]
}

File: python/mirage/commands/builtin/generic/crossmount/fanout/fanout.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.commands.builtin.generic.crossmount.fanout.du import du_total
from mirage.commands.builtin.generic.crossmount.fanout.exit import \
    combined_exit
from mirage.commands.builtin.generic.crossmount.fanout.wc import combine_wc
from mirage.commands.builtin.generic.crossmount.types import (Cmd, CrossResult,
                                                              RunSingle)
from mirage.commands.builtin.generic.crossmount.utils import (
    merge_operand_ios, run_operands)
from mirage.commands.spec import SPECS
from mirage.commands.spec.types import FlagView
from mirage.io.stream import materialize
from mirage.io.types import ByteSource
from mirage.types import PathSpec


async def run_fanout(cmd_name: str,
                     scopes: list[PathSpec],
                     text_args: list[str],
                     flag_kwargs: dict[str, object],
                     run_single: RunSingle,
                     stdin: ByteSource | None = None) -> CrossResult:
    """Run a per-operand command whose operands span mounts.

    The command runs natively once per operand on the operand's owning
    mount (globs expand inside that native run), and the outputs combine
    in operand order. Filename-keyed commands stay correct because every
    native run is forced to name its files (grep ``-H``, head/tail ``-v``);
    wc and ``du -c`` re-total across runs.

    Args:
        cmd_name (str): One of the FANOUT_COMMANDS (or ``sed -i``).
        scopes (list[PathSpec]): Path operands in command-line order.
        text_args (list[str]): Positional text operands (grep pattern,
            find expression).
        flag_kwargs (dict): Flags parsed against the shared command spec.
        run_single (RunSingle): Executor-injected single-mount runner.
        stdin (ByteSource | None): Original stdin, re-fed per operand (tee).
    """
    flags = dict(flag_kwargs)
    stdin_bytes: bytes | None = None
    if cmd_name == Cmd.TEE:
        stdin_bytes = await materialize(stdin) if stdin is not None else b""
    if cmd_name == Cmd.GREP and not FlagView(
            flags, spec=SPECS[Cmd.GREP]).as_bool("h"):
        flags["H"] = True
    if cmd_name == Cmd.RG and not FlagView(
            flags, spec=SPECS[Cmd.RG]).as_bool("args_I"):
        flags["H"] = True
    if cmd_name in (Cmd.HEAD, Cmd.TAIL) and not FlagView(
            flags, spec=SPECS[cmd_name]).as_bool("q"):
        flags["v"] = True

    results = await run_operands(run_single,
                                 cmd_name,
                                 scopes,
                                 list(text_args),
                                 flags,
                                 stdin_bytes=stdin_bytes)
    errored = [
        r.io.exit_code != 0 and r.io.stderr is not None for r in results
    ]
    quiet = cmd_name == Cmd.GREP and FlagView(
        flags, spec=SPECS[Cmd.GREP]).as_bool("q")
    exit_code = combined_exit(cmd_name, [r.io.exit_code for r in results],
                              errored, quiet)

    if cmd_name == Cmd.WC:
        body = combine_wc(results, flag_kwargs)
    elif cmd_name == Cmd.DU and FlagView(flag_kwargs,
                                         spec=SPECS[Cmd.DU]).as_bool("c"):
        body = du_total(results,
                        FlagView(flag_kwargs, spec=SPECS[Cmd.DU]).as_bool("h"))
    elif cmd_name == Cmd.TEE:
        body = stdin_bytes or b""
    elif (cmd_name in (Cmd.HEAD, Cmd.TAIL)
          and FlagView(flags, spec=SPECS[cmd_name]).as_bool("v")) or (
              cmd_name == Cmd.LS
              and FlagView(flags, spec=SPECS[Cmd.LS]).as_bool("R")):
        # Blank line between per-operand blocks, like one native run
        # separates its own file blocks.
        body = b"\n".join(r.data for r in results if r.data)
    else:
        body = b"".join(r.data for r in results)

    io = await merge_operand_ios(results, exit_code)
    return body, io
